Elasticsearch scroll is created even when there is no scroll v0.5.0

Please post all questions and issues on [janusgraph-users](https://groups.google.com/forum/#!forum/janusgraph-users)
before opening a GitHub issue. Your questions will reach a wider audience there,
and if we confirm that there is a bug, then you can open a new issue.

Please include configurations and logs if available.

For confirmed bugs, please report:
- Version: 0.5.0
- Mixed Index Backend: ElasticSearch
- Mailing list Thread URL: https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!msg/janusgraph-users/Bv_tYFI38kc/QJIYrCgYAgAJ
- Steps to Reproduce: 

Access mixedIndex in ElasticSearch with limit which will be less then the batch limit. In such case the scroll will not be used but we always create `ElasticSearchScroll` for results even when scroll if not used at all. In such case, `ElasticSearchScroll` will close scroll immediately and the error which be raised (because there is not any scrolls).

### Stack Trace (if you have one)

```
[gremlin-server-exec-37] WARN  org.apache.tinkerpop.gremlin.server.op.traversal.TraversalOpProcessor  - Exception processing a Traversal on iteration for request [d8f41122-74a4-4bfc-9c81-58b53011545d].
org.janusgraph.core.JanusGraphException: Could not call index
	at org.janusgraph.graphdb.util.SubqueryIterator.<init>(SubqueryIterator.java:68)
	at org.janusgraph.graphdb.transaction.StandardJanusGraphTx$3.execute(StandardJanusGraphTx.java:1341)
	at org.janusgraph.graphdb.transaction.StandardJanusGraphTx$3.execute(StandardJanusGraphTx.java:1233)
	at org.janusgraph.graphdb.query.QueryProcessor$LimitAdjustingIterator.getNewIterator(QueryProcessor.java:194)
	at org.janusgraph.graphdb.query.LimitAdjustingIterator.hasNext(LimitAdjustingIterator.java:68)
	at com.google.common.collect.Iterators$7.computeNext(Iterators.java:650)
	at com.google.common.collect.AbstractIterator.tryToComputeNext(AbstractIterator.java:143)
	at com.google.common.collect.AbstractIterator.hasNext(AbstractIterator.java:138)
	at org.janusgraph.graphdb.query.ResultSetIterator.nextInternal(ResultSetIterator.java:54)
	at org.janusgraph.graphdb.query.ResultSetIterator.<init>(ResultSetIterator.java:44)
	at org.janusgraph.graphdb.query.QueryProcessor.iterator(QueryProcessor.java:66)
	at com.google.common.collect.Iterables$7.iterator(Iterables.java:613)
	at org.janusgraph.graphdb.tinkerpop.optimize.JanusGraphStep.executeGraphCentricQuery(JanusGraphStep.java:155)
	at org.janusgraph.graphdb.tinkerpop.optimize.JanusGraphStep.lambda$null$1(JanusGraphStep.java:94)
	at java.lang.Iterable.forEach(Iterable.java:75)
	at org.janusgraph.graphdb.tinkerpop.optimize.JanusGraphStep.lambda$new$2(JanusGraphStep.java:94)
	at org.apache.tinkerpop.gremlin.process.traversal.step.map.GraphStep.processNextStart(GraphStep.java:157)
	at org.apache.tinkerpop.gremlin.process.traversal.step.util.AbstractStep.hasNext(AbstractStep.java:143)
	at org.apache.tinkerpop.gremlin.process.traversal.step.util.ExpandableStepIterator.next(ExpandableStepIterator.java:50)
	at org.apache.tinkerpop.gremlin.process.traversal.step.filter.FilterStep.processNextStart(FilterStep.java:37)
	at org.apache.tinkerpop.gremlin.process.traversal.step.util.AbstractStep.hasNext(AbstractStep.java:143)
	at org.apache.tinkerpop.gremlin.process.traversal.step.util.ExpandableStepIterator.hasNext(ExpandableStepIterator.java:42)
	at org.apache.tinkerpop.gremlin.process.traversal.step.util.ReducingBarrierStep.processAllStarts(ReducingBarrierStep.java:82)
	at org.apache.tinkerpop.gremlin.process.traversal.step.util.ReducingBarrierStep.processNextStart(ReducingBarrierStep.java:112)
	at org.apache.tinkerpop.gremlin.process.traversal.step.util.AbstractStep.hasNext(AbstractStep.java:143)
	at org.apache.tinkerpop.gremlin.process.traversal.step.util.ExpandableStepIterator.next(ExpandableStepIterator.java:50)
	at org.apache.tinkerpop.gremlin.process.traversal.step.map.FlatMapStep.processNextStart(FlatMapStep.java:48)
	at org.apache.tinkerpop.gremlin.process.traversal.step.util.AbstractStep.hasNext(AbstractStep.java:143)
	at org.apache.tinkerpop.gremlin.process.traversal.step.util.ExpandableStepIterator.next(ExpandableStepIterator.java:50)
	at org.apache.tinkerpop.gremlin.process.traversal.step.map.MapStep.processNextStart(MapStep.java:36)
	at org.apache.tinkerpop.gremlin.process.traversal.step.util.AbstractStep.hasNext(AbstractStep.java:143)
	at org.apache.tinkerpop.gremlin.process.traversal.util.DefaultTraversal.hasNext(DefaultTraversal.java:197)
	at org.apache.tinkerpop.gremlin.server.util.TraverserIterator.fillBulker(TraverserIterator.java:69)
	at org.apache.tinkerpop.gremlin.server.util.TraverserIterator.hasNext(TraverserIterator.java:56)
	at org.apache.tinkerpop.gremlin.server.op.traversal.TraversalOpProcessor.handleIterator(TraversalOpProcessor.java:512)
	at org.apache.tinkerpop.gremlin.server.op.traversal.TraversalOpProcessor.lambda$iterateBytecodeTraversal$4(TraversalOpProcessor.java:411)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: org.janusgraph.diskstorage.PermanentBackendException: Permanent failure in storage backend
	at org.janusgraph.diskstorage.es.ElasticSearchIndex.query(ElasticSearchIndex.java:1120)
	at org.janusgraph.diskstorage.indexing.IndexTransaction.queryStream(IndexTransaction.java:108)
	at org.janusgraph.diskstorage.BackendTransaction$6.call(BackendTransaction.java:416)
	at org.janusgraph.diskstorage.BackendTransaction$6.call(BackendTransaction.java:413)
	at org.janusgraph.diskstorage.util.BackendOperation.executeDirect(BackendOperation.java:68)
	at org.janusgraph.diskstorage.util.BackendOperation.execute(BackendOperation.java:54)
	at org.janusgraph.diskstorage.BackendTransaction.executeRead(BackendTransaction.java:469)
	at org.janusgraph.diskstorage.BackendTransaction.indexQuery(BackendTransaction.java:413)
	at org.janusgraph.graphdb.database.IndexSerializer.query(IndexSerializer.java:533)
	at org.janusgraph.graphdb.util.SubqueryIterator.<init>(SubqueryIterator.java:66)
	... 41 more
Caused by: java.io.UncheckedIOException: method [DELETE], host [http://11.182.230.151:9200], URI [/_search/scroll/null], status line [HTTP/1.1 400 Bad Request]
{"error":{"root_cause":[{"type":"illegal_argument_exception","reason":"Cannot parse scroll id"}],"type":"illegal_argument_exception","reason":"Cannot parse scroll id","caused_by":{"type":"array_index_out_of_bounds_exception","reason":null}},"status":400}
	at org.janusgraph.diskstorage.es.ElasticSearchScroll.update(ElasticSearchScroll.java:52)
	at org.janusgraph.diskstorage.es.ElasticSearchScroll.<init>(ElasticSearchScroll.java:43)
	at org.janusgraph.diskstorage.es.ElasticSearchIndex.query(ElasticSearchIndex.java:1115)
	... 50 more
Caused by: org.elasticsearch.client.ResponseException: method [DELETE], host [http://11.182.230.151:9200], URI [/_search/scroll/null], status line [HTTP/1.1 400 Bad Request]
{"error":{"root_cause":[{"type":"illegal_argument_exception","reason":"Cannot parse scroll id"}],"type":"illegal_argument_exception","reason":"Cannot parse scroll id","caused_by":{"type":"array_index_out_of_bounds_exception","reason":null}},"status":400}
	at org.elasticsearch.client.RestClient.convertResponse(RestClient.java:283)
	at org.elasticsearch.client.RestClient.performRequest(RestClient.java:261)
	at org.elasticsearch.client.RestClient.performRequest(RestClient.java:235)
	at org.janusgraph.diskstorage.es.rest.RestElasticSearchClient.deleteScroll(RestElasticSearchClient.java:450)
	at org.janusgraph.diskstorage.es.ElasticSearchScroll.update(ElasticSearchScroll.java:50)
	... 52 more
```

--- a/janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphBaseTest.java
+++ b/janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphBaseTest.java
@@ -453,13 +453,17 @@ public abstract class JanusGraphBaseTest implements JanusGraphBaseStoreFeaturesT
         return StreamSupport.stream(iterable.spliterator(),false);
     }
 
-    public JanusGraph getForceIndexGraph() throws BackendException {
-        final ModifiableConfiguration adjustedConfig = new ModifiableConfiguration(GraphDatabaseConfiguration.ROOT_NS, getConfiguration(), BasicConfiguration.Restriction.NONE);
+    public JanusGraph getForceIndexGraph() {
+        return getForceIndexGraph(getConfiguration());
+    }
+
+    public JanusGraph getForceIndexGraph(WriteConfiguration writeConfiguration) {
+        final ModifiableConfiguration adjustedConfig = new ModifiableConfiguration(GraphDatabaseConfiguration.ROOT_NS, writeConfiguration, BasicConfiguration.Restriction.NONE);
         adjustedConfig.set(GraphDatabaseConfiguration.FORCE_INDEX_USAGE, true);
-        final WriteConfiguration writeConfig = adjustedConfig.getConfiguration();
-        TestGraphConfigs.applyOverrides(writeConfig);
-        Preconditions.checkNotNull(writeConfig);
-        return JanusGraphFactory.open(writeConfig);
+        final WriteConfiguration adjustedWriteConfig = adjustedConfig.getConfiguration();
+        TestGraphConfigs.applyOverrides(adjustedWriteConfig);
+        Preconditions.checkNotNull(adjustedWriteConfig);
+        return JanusGraphFactory.open(adjustedWriteConfig);
     }
 
 }

--- a/janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphIndexTest.java
+++ b/janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphIndexTest.java
@@ -2225,6 +2225,24 @@ public abstract class JanusGraphIndexTest extends JanusGraphBaseTest {
     }
 
     @Test
+    public void testIndexDataRetrievalWithLimitLessThenBatch() throws Exception {
+        WriteConfiguration config = getConfiguration();
+        config.set("index.search.max-result-set-size", 10);
+        JanusGraph customGraph = getForceIndexGraph(config);
+        final JanusGraphManagement management = customGraph.openManagement();
+        final PropertyKey num = management.makePropertyKey("num").dataType(Integer.class).cardinality(Cardinality.SINGLE).make();
+        management.buildIndex("oridx", Vertex.class).addKey(num).buildMixedIndex(INDEX);
+        management.commit();
+        customGraph.tx().commit();
+        final GraphTraversalSource g = customGraph.traversal();
+        g.addV().property("num", 1).next();
+        g.addV().property("num", 2).next();
+        customGraph.tx().commit();
+        assertEquals(2, customGraph.traversal().V().has("num", P.lt(3)).limit(4).toList().size());
+        JanusGraphFactory.close(customGraph);
+    }
+
+    @Test
     public void testOrForceIndexComposite() throws Exception {
         JanusGraph customGraph = null;
         try {

--- a/janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/ElasticSearchIndex.java
+++ b/janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/ElasticSearchIndex.java
@@ -1112,13 +1112,17 @@ public class ElasticSearchIndex implements IndexProvider {
                 compat.createRequestBody(sr, useScroll? NULL_PARAMETERS : TRACK_TOTAL_HITS_DISABLED_PARAMETERS),
                 useScroll);
             log.debug("First Executed query [{}] in {} ms", query.getCondition(), response.getTook());
-            final ElasticSearchScroll resultIterator = new ElasticSearchScroll(client, response, sr.getSize());
+            final Iterator<RawQuery.Result<String>> resultIterator = getResultsIterator(useScroll, response, sr.getSize());
             final Stream<RawQuery.Result<String>> toReturn
                     = StreamSupport.stream(Spliterators.spliteratorUnknownSize(resultIterator, Spliterator.ORDERED), false);
             return (query.hasLimit() ? toReturn.limit(query.getLimit()) : toReturn).map(RawQuery.Result::getResult);
         } catch (final IOException | UncheckedIOException e) {
             throw new PermanentBackendException(e);
         }
+    }
+
+    private Iterator<RawQuery.Result<String>> getResultsIterator(boolean useScroll, ElasticSearchResponse response, int windowSize){
+        return (useScroll)? new ElasticSearchScroll(client, response, windowSize) : response.getResults().iterator();
     }
 
     private String convertToEsDataType(Class<?> dataType, Mapping mapping) {
@@ -1206,9 +1210,10 @@ public class ElasticSearchIndex implements IndexProvider {
     public Stream<RawQuery.Result<String>> query(RawQuery query, KeyInformation.IndexRetriever information,
                                                  BaseTransaction tx) throws BackendException {
         final int size = query.hasLimit() ? Math.min(query.getLimit() + query.getOffset(), batchSize) : batchSize;
-        final ElasticSearchResponse response = runCommonQuery(query, information, tx, size, size >= batchSize );
+        final boolean useScroll = size >= batchSize;
+        final ElasticSearchResponse response = runCommonQuery(query, information, tx, size, useScroll);
         log.debug("First Executed query [{}] in {} ms", query.getQuery(), response.getTook());
-        final ElasticSearchScroll resultIterator = new ElasticSearchScroll(client, response, size);
+        final Iterator<RawQuery.Result<String>> resultIterator = getResultsIterator(useScroll, response, size);
         final Stream<RawQuery.Result<String>> toReturn
                 = StreamSupport.stream(Spliterators.spliteratorUnknownSize(resultIterator, Spliterator.ORDERED),
                 false).skip(query.getOffset());

